File: Labs/week7/C_complete_binary_tree_level.cpp
#include <iostream>
using namespace std;

const int N = 2000001;
int a[N], post[N], mp[N], idx = 0, n;

void postorder(int x) {
    if (x > n) return;
    postorder(x * 2);
    postorder(x * 2 + 1);
    post[++idx] = x;
}

int main() {
    cin >> n;
    for (int i = 1; i <= n; i++) cin >> a[i];
    postorder(1);
    for (int i = 1; i <= n; i++) mp[post[i]] = a[i];
    for (int i = 1; i <= n; i++) {
        cout << mp[i];
        if (i == n) cout << endl;
        else cout << " ";
    }
    return 0;
}

File: Labs/week7/C_complete_binary_tree_level.py
n = int(input())
a = [0] + list(map(int, input().split()))
post = []
idx = 0

def postorder(x):
    global idx
    if x > n:
        return
    postorder(x * 2)
    postorder(x * 2 + 1)
    post.append(x)

postorder(1)
mp = [0] * (n + 1)
for i in range(1, n + 1):
    mp[post[i - 1]] = a[i]

print(' '.join(str(mp[i]) for i in range(1, n + 1)))
